### PR TITLE
TWEAK/FEATURE: Battle-Train UI/System Overhaul +Saves Backups Feature

### PR DESCRIPTION
**Pull request by OcelotWalrus**
# Tweak

This PR addresses the game progression goal #'Enhance battle fights'

## Summary
Here's the changes of this PR:
- Update both game version from 0.17 to 0.18
- A new checks when loading save, to determine if the save file's corrupted. If it is, it'll throwback to cout a message to say that the save 'x' is corrupted and to check the logs for further information
- Balance the defend formula: instead of being `defend += random.randint(0, int(player_fake_defend)) * player_fake_agility`, it's now `defend += random.randint(int(player_fake_defend / 2), int(player_fake_defend)) * player_fake_agility`
- Balance the dodge formula, by adding a 35% chance to the base formula
- Fix some tyopes when printing stuff
- Fix some unneeded imports
- Fixed bugs with the `logger_sys.py`
- Fixed bugs happening when the player died
- Added the `$spawn$enemy$` secret command, that spawns a custom enemy(ies)
- Completely overhaul the in-fight UI
- Completely overhaul the training UI
- Add the ability to load and create backups of saves files

## Screenshots
<details>

<summary>Shots</summary>

__The `$spawn$enemy$` command:__

image

__New in-fight UI:__

Screenshot from 2024-03-11 18-21-01
Screenshot from 2024-03-11 18-21-08
Screenshot from 2024-03-11 18-21-21

__New train UI:__

image

</details>

## Testing Done
Everything's tested when coded.

## Save File
It's recommended to just create a new save to test the changes.

## Performance Impact
Some, because of some checks. But it isn't noticeable by players.

## Check-List
- [x] Update game version
- [x] Add checks to determine if a loaded save file is corrupted
- [x] Balance battle formulas
- [x] Overhaul the in-fight UI
- [x] Overhaul training UI
- [x] Add load/create save backup feature
- [x] Update the wiki
- [x] Fix code style checks

### CODE REVIEW

File: source/battle.py [hunk 1]
```diff
@@ -538,23 +601,15 @@ def fight(
...
-                    cout(
-                        f"|{health_color_enemy}{remaining_health_bars_enemy * remaining_health_symbol}" +
-                        f"{lost_health_bars_enemy * lost_health_symbol}{color_default}|"
-                    )
                     player["xp"] += enemy_max * enemy_max_damage / 3
                     if player["current mount"] in player["mounts"]:
                         player["mounts"][player["current mount"]]["level"] += round(random.uniform(.05, .20), 3)
                     player["health"] += random.randint(0, 3)
                     enemies_remaining -= 1
+                    cout("You killed one {enemy_singular}!")
```
Comment (lumbar527) [inline source/battle.py:609]:
```suggestion
                    cout("You killed text_handling.a_an_check(enemy_singular)!")
```
Just a small suggestion

File: source/main.py [hunk 2]
```diff
@@ -287,9 +288,30 @@ def print_title():
...
+                    error_loading = True
+                previous_player = player
                 check_yaml.examine(save_file)
+                if type(player) is not type({}) and not error_loading:
+                    error_loading = True
+                    error = 'not a yaml file'
+                if error_loading:
+                    cout(
+                        COLOR_RED + "FATAL ERROR: " + COLOR_STYLE_BRIGHT +
+                        "save corrupted! Check logs files for further information" + COLOR_RESET_ALL
```
Comment (lumbar527) [inline source/main.py:304]:
```suggestion
                        "Save corrupted! Check logs files for further information" + COLOR_RESET_ALL
```

File: source/main.py [hunk 3]
```diff
@@ -287,9 +288,30 @@ def print_title():
...
+                    )
+                    logger_sys.log_message(f"FATAL ERROR: save '{save_file}' corrupted!")
```
Comment (lumbar527) [inline source/main.py:306]:
```suggestion
                    logger_sys.log_message(f"FATAL ERROR: Save '{save_file}' corrupted!")
```

File: source/main.py [hunk 4]
```diff
@@ -366,9 +388,30 @@ def print_title():
...
+                        error_loading = True
+                    previous_player = player
                     check_yaml.examine(save_file)
+                    if type(player) is not type({}) and not error_loading:
+                        error_loading = True
+                        error = 'not a yaml file'
+                    if error_loading:
+                        cout(
+                            COLOR_RED + "FATAL ERROR: " + COLOR_STYLE_BRIGHT +
+                            "save corrupted! Check logs files for further information" + COLOR_RESET_ALL
```
Comment (lumbar527) [inline source/main.py:404]:
```suggestion
                            "Save corrupted! Check logs files for further information" + COLOR_RESET_ALL
```

File: source/main.py [hunk 5]
```diff
@@ -366,9 +388,30 @@ def print_title():
...
+                        )
+                        logger_sys.log_message(f"FATAL ERROR: save '{save_file}' corrupted!")
```
Comment (lumbar527) [inline source/main.py:406]:
```suggestion
                        logger_sys.log_message(f"FATAL ERROR: Save '{save_file}' corrupted!")
```

File: source/main.py [hunk 6]
```diff
@@ -398,8 +441,29 @@ def print_title():
...
+                    except Exception as error:
+                        error_loading = True
+                    previous_player = player
+                    if type(player) is not type({}) and not error_loading:
+                        error_loading = True
+                        error = 'not a yaml file'
+                    if error_loading:
+                        cout(
+                            COLOR_RED + "FATAL ERROR: " + COLOR_STYLE_BRIGHT +
+                            "save corrupted! Check logs files for further information" + COLOR_RESET_ALL
```
Comment (lumbar527) [inline source/main.py:456]:
```suggestion
                            "Save corrupted! Check logs files for further information" + COLOR_RESET_ALL
```

File: source/main.py [hunk 7]
```diff
@@ -398,8 +441,29 @@ def print_title():
...
+                        )
+                        logger_sys.log_message(f"FATAL ERROR: save '{save_file}' corrupted!")
```
Comment (lumbar527) [inline source/main.py:458]:
```suggestion
                        logger_sys.log_message(f"FATAL ERROR: Save '{save_file}' corrupted!")
```

File: source/main.py [hunk 8]
```diff
@@ -2865,8 +3009,29 @@ def run(play):
...
+                    except Exception as error:
+                        error_loading = True
+                    previous_player = player
+                    if type(player) is not type({}) and not error_loading:
+                        error_loading = True
+                        error = 'not a yaml file'
+                    if error_loading:
+                        cout(
+                            COLOR_RED + "FATAL ERROR: " + COLOR_STYLE_BRIGHT +
+                            "save corrupted! Check logs files for further information" + COLOR_RESET_ALL
```
Comment (lumbar527) [inline source/main.py:3024]:
```suggestion
                            "Save corrupted! Check logs files for further information" + COLOR_RESET_ALL
```

File: source/main.py [hunk 9]
```diff
@@ -2865,8 +3009,29 @@ def run(play):
...
+                        )
+                        logger_sys.log_message(f"FATAL ERROR: save '{save_file}' corrupted!")
```
Comment (lumbar527) [inline source/main.py:3026]:
```suggestion
                        logger_sys.log_message(f"FATAL ERROR: Save '{save_file}' corrupted!")
```

**Review by lumbar527 — COMMENTED** [review]
Mostly some capitalization suggestions.
Nice to see you're still working on this!

File: source/battle.py (hunk 1 above)
Comment (OcelotWalrus) [reply source/battle.py:609]: You've forgotten to put the function in '{' brackets! Could you edit that so I commit it then?

**Comment by OcelotWalrus** [comment]
Hey @lumbar527! Nice to see you! Thanks for your review, I really appreciate it!

File: source/battle.py (hunk 1 above)
Comment (lumbar527) [reply source/battle.py:609]: `cout()` works like print then?
I fixed my suggestion.

Comment (OcelotWalrus) [reply source/battle.py:609]: > `cout()` works like print then? I fixed my suggestion.

Yes, it's just a custom function to print stuff.

**Comment by OcelotWalrus** [comment]
Hey @lumbar527, any other suggestions?

File: source/train.py [hunk 10]
```diff
@@ -90,13 +139,14 @@ def training_loop(mount_uuid, player, item, mounts, stable, time_elapsing_coeffi
                 cout(COLOR_YELLOW, end="")
                 text_handling.print_long_string(text)
                 cout(COLOR_RESET_ALL, end="")
-        elif choice == 'Train':
+        elif choice.startswith('t'):
```
Comment (lumbar527) [inline source/train.py:142]:
```suggestion
        elif choice.lower().startswith('t'):
```

**Review by lumbar527 — APPROVED** [review]
Not much to add, I haven't tested though.
The changes sound good I think.

File: source/train.py (hunk 10 above)
Comment (OcelotWalrus) [reply source/train.py:142]: Actually the .lower() function is already applied to the choice variable:
```python
choice = cinput(COLOR_GREEN + COLOR_STYLE_BRIGHT + "> " + COLOR_RESET_ALL).lower()
```

File: source/battle.py [hunk 11]
```diff
@@ -377,56 +386,100 @@ def fight(
...
+
+            # print HP stats and possible actions for the player
+
+            if risk > .80 * 100:
+                health_color = COLOR_RED
+            elif risk > .60 * 100:
+                health_color = COLOR_ORANGE_4
+            elif risk > .45 * 100:
+                health_color = COLOR_YELLOW
+            elif risk > .30 * 100:
```
[the remark this answers is not in the document]
Comment (OcelotWalrus) [reply source/battle.py:420]: It's not really necessary to do that.

**Pull request merged**